src/dune_cache_storage/layout.ml

```diff
@@ -16,28 +16,68 @@ let root_path =
 
 let ( / ) = Path.relative
 
-(* We version metadata and actual cache content separately. *)
-let metadata_storage_path = root_path / "meta" / "v5"
-
-let file_storage_path = root_path / "files" / "v4"
-
-let value_storage_path = root_path / "values" / "v3"
+let temp_path = root_path / "temp"
 
 let cache_path ~dir ~hex =
   let two_first_chars = sprintf "%c%c" hex.[0] hex.[1] in
   dir / two_first_chars / hex
 
-let metadata_path ~rule_or_action_digest =
-  cache_path ~dir:metadata_storage_path
-    ~hex:(Digest.to_string rule_or_action_digest)
+(* List all entries in a given storage directory. *)
+let list_entries ~storage =
+  let open Result.O in
+  let entries dir =
+    match
+      String.length dir = 2 && String.for_all ~f:Char.is_lowercase_hex dir
+    with
+    | false ->
+      (* Ignore directories whose name isn't a two-character hex value. *)
+      Ok []
+    | true ->
+      let dir = storage / dir in
+      Path.readdir_unsorted dir >>| List.map ~f:(Path.relative dir)
```

Comment:
Given that we skip unknown directories at the first level, we could also skip unknown directories at the second level. i.e. ignore anything that is not `<2-hex-digits>/<32-hex-digits>`. What do you think?

Reply:
Agreed. As you suggested before, I'll make the output of this function more structured, returning the digests digests instead of just paths, and I'll skip all files whose names are not digests.

Reply:
Done.